elastic/beats
# Add drop and explicit tests to avoid duplicate ingest of elasticsearch logs

## What does this PR do?

Adds a "drop" to the elasticsearch pipeline as well as explicit "mixed" test logs to ensure we won't attempt to ingest logs across mismatched pipelines.

## Why is it important?

Without it, the elasticsearch slowlog pipeline will attempt to ingest all 5 file sets shipped by the elasticsearch filebeat module.

Additionally the test cases help guard against removal of the `drop` processors.

## Checklist

- ~~I have made corresponding changes to the documentation~~
- ~~I have made corresponding change to the default configuration files~~

## How to test this PR locally

See https://www.elastic.co/guide/en/beats/devguide/master/filebeat-modules-devguide.html#_test for setup instructions. Specify `TESTING_FILEBEAT_MODULES=elasticsearch` to test the elasticsearch module directly.

## Related issues

Fixes https://github.com/elastic/beats/issues/30428
Related https://github.com/elastic/beats/pull/30164
Related https://github.com/elastic/beats/issues/16540

## Use cases

In Elasticsearch log4j2.properties config is defined that server, deprecation, slowlog and audit logs are written to Console:

```
grep -R "Console" log4j2.properties
appender.rolling.type = Console
appender.deprecation_rolling.type = Console
appender.index_search_slowlog_rolling.type = Console
appender.index_indexing_slowlog_rolling.type = Console
appender.audit_rolling.type = Console
```

On the kubernetes node, where elasticsearch container is running, all those logs will be in one file:

/var/log/containers/<elasticsearch-pod-name>_<elasticsearch-namespace>_elasticsearch-<elasticsearch-container-id>.log

Filebeat pod has the whole folder `/var/log ` mounted and reads log files from `/var/log/containers/`.

The elasticsearch module has 5 filesets which will lead to reading the kubernetes log 5 times and shipping every message to each pipeline.

This works around the issue of duplicate log storage by dropping at the top of the pipeline.
